Document new programmatic APIs and options (#11051)
withastro/docs: src/content/docs/en/reference/programmatic-reference.mdx

await sync({
  root: "./my-project",
});
```

## `mergeConfig()`

<p>

**Type:** `<T extends AstroConfig | AstroInlineConfig>(config: T, overrides: DeepPartial<T>) => T`
</p>

Imported from `astro/config`, merges a partial Astro configuration on top of an existing, valid, Astro configuration.

> 

Reviewer: ```suggestion

The merge behaves according to the following rules:

```
Do we need some kind of sentence to introduce the bullet points maybe? (And, maybe that's not the right sentence -- just off the top of my head. It seems like we should set some context for "list of bullet points.")

> - General options are overriden.

Reviewer: ```suggestion
- General options in the existing config are overriden.
```
(Also, I'm not sure what "general" options are? Is that a specific thing? If not, is there a proper name for these options? "top-level"?)

Author: Anything that doesn't match one of the other rules. Even if they are nested.

Reviewer: OK, then should this point come *last*? As in:

- All other options override the existing config

- Integrations and remark plugins are appended to the existing list.

> - Vite options are merged using [Vite's own `mergeConfig` function](https://vite.dev/guide/api-javascript#mergeconfig) with default `isRoot` flag.

Reviewer: ```suggestion
- Vite options are merged using [Vite's own `mergeConfig` function](https://vite.dev/guide/api-javascript#mergeconfig) with the default `isRoot` flag.
```

> - Lazy options passed as functions are wrapped into new functions that recursively merge the return values from both configurations with these same rules.

Reviewer: Similar comment here, is "lazy options" something that is generally understood as a thing? I'm assuming this means "passing a function to set an option, which therefore needs to be worked into a process" vs "passing an option directly by providing a new config object to merge with the existing one"?

I'm trying to figure out the exact parsing of this to know how best to evaluate this line. Is it... ?

- Lazy options (WHICH MEANS passed as functions which will set a config value vs set one directly)
- Lazy options THAT HAPPEN TO BE PASSED AS FUNCTIONS (but they could totally be passed a different way)
- something entirely else?

Author: First option. Some options allow passing either the value or a function returning the value. For example:

```ts
defineConfig({
  server: {
    headers: {
      foo: 'bar',
    },
  },
})

defineConfig({
  server: ({command}) => ({
    headers: {
      astroCommand: command,
    },
  }),
})
```

Reviewer: Great, thanks for the clarification! Then maybe something like one of these just to avoid the ambiguous parsing...

- Lazy options (passed as functions) are wrapped into new functions ... (if you want to emphasize the term "lazy options" because that's particularly meaningful)
- Options configured by passing functions are wrapped into new functions ... (if you don't care about "lazy" and instead want to describe what the config looks like)

Also just want to check on the different "effect" or end result of this vs. just passing a value directly:
- is there an example of a situation where passing a function results in something different than overriding an existing config with a directly passed value?
- How is knowing that this process creates a new, recursive function helpful to the reader? Would they do anything differently because they know this? (In other words, is this more a "fun fact" / implementation detail than something people need to know to make sure they merge configs properly?
- Are there any particular gotchas/caveats/warnings about passing a config option this way that people should be aware of? (Other than I guess not all options support this method?) Does this "slow down" the process, or resolve itself in ways that might conflict with anything else?

Author: > How is knowing that this process creates a new, recursive function helpful to the reader? Would they do anything differently because they know this? (In other words, is this more a "fun fact" / implementation detail than something people need to know to make sure they merge configs properly?

Most "merge" functions in the wild would just replace the function with the new one instead of merging the result. That is the behavior of a recursive use of the spread operator (`{...current, ...overrides}`)

> ```ts

Reviewer: Can we have a line explaining what this example is supposed to be showing? We've talked about a bunch of different things above including passing a function. Should we also show an example of a plugin or even function being passed here, or is that going to look too complicated?

Something like: 

The following example uses `mergeConfig` to append first, Astro's Tailwind and then MDX integrations to an existing config.

(I'm inferring from the example that if you have multiple integrations to add, you have to add multiple integrations separately? You can't combine it into one `integrations: []` array? This is maybe worth mentioning somewhere, too, that every single new thing you want to merge into an existing config must be passed as a separate (top-level?) object?)

Seeing this all now, I'm also tempted to reorder the entry as:

```
## `mergeConfig()`

<p>Type stuff</p>

Imported from `astro/config`, merges a partial Astro configuration on top of an existing, valid, Astro configuration.

Code Example so people just *see* it immediately

Each config value you wish to add or overwrite must be added as its own, individual config object.
For example, multiple integrations must be added separately, in the order in which you want to append them to the existing config. 

(ASIDE: what happens if you want an integration added earlier?? Does this ever happen/matter??)

Merging behavior follows these rules + list of bullet points
```

Author: I'll update the example to show a case with every item.

import { mergeConfig } from "astro/config";

const config = mergeConfig(
  {
    integrations: [tailwind()],
  },
  {
    integrations: [mdx()],
  }
);
```

## `validateConfig()`

<p>

**Type:** `(userConfig: any, root: string, cmd: string): Promise<AstroConfig>`
</p>

Imported from `astro/config`, validates an object as if it was exported from `astro.config.mjs` and imported by Astro.

It takes the following arguments: 
- The configuration to be validated.
- The root directory of the project.
- The Astro command that is being executed (`build`, `dev`, `sync`, etc.)

The returned promise resolves to the validated configuration, filled with all default values appropriate for the given Astro command.

> ```ts

Reviewer: Noting that unlike the previous entry, I'm not going to suggest re-ordering here to put the code sample earlier because unlike before, the earlier text explains arguments, return value etc.

Previously, the additional text wasn't as much *what you absolutely needed to know before writing the code example* but a subsequent description of how that was going to work.

This feels like the correct order here, and if that means you'd prefer not to reorder above to have them match, that's fine.

(But, if I were looking at each one on its own, `mergeConfig()` feels like something where I'd naturally put the example earlier, because it's easier to grok just by looking at it, whereas someone looking up `validateConfig()` probably is checking "what are all the arguments? etc." and needs that info sooner.)

await sync({
  root: "./my-project",
});
```

## `mergeConfig()`

<p>

**Type:** `<T extends AstroConfig | AstroInlineConfig>(config: T, overrides: DeepPartial<T>) => T`
</p>

Imported from `astro/config`, merges a partial Astro configuration on top of an existing, valid, Astro configuration.

> The return value is a valid Astro config combining both values such that:

Reviewer: ```suggestion
`mergeConfig()` accepts two Astro config objects, and the return value is a valid Astro config combining both values such that:
```
Is this correct to say? At first, I didn't realize you needed to give it two different configs to merge; I thought it was "merge whatever you already have with this new thing I'm giving you", so it's maybe helpful to say this out in words (not just in the type declaration?)

Author: The first argument is a config, the second is any part of a config, everything is optional on it (that is the `DeepPartial` in the types). Sending another option!

- Arrays are concatenated (including integrations and remark plugins).
- Objects are merged recursively.
- Vite options are merged using [Vite's own `mergeConfig` function](https://vite.dev/guide/api-javascript#mergeconfig) with the default `isRoot` flag.
- Options that can be provided as functions are wrapped into new functions that recursively merge the return values from both configurations with these same rules.
- All other options override the existing config.

> ```ts

Reviewer: This example looks great!

await sync({
  root: "./my-project",
});
```

## `mergeConfig()`

<p>

**Type:** `<T extends AstroConfig | AstroInlineConfig>(config: T, overrides: DeepPartial<T>) => T`
<Since v="5.4.0" />
</p>

Imported from `astro/config`, merges a partial Astro configuration on top of an existing, valid, Astro configuration.

> `mergeConfig()` accepts an Astro config object and a partial config (any portion of a config, everything is optional), and returns a valid Astro config combining both values such that:

Reviewer: ```suggestion
`mergeConfig()` accepts an Astro config object and a partial config (any set of valid Astro config options), and returns a valid Astro config combining both values such that:
```
I like it! If I tweak it a tiny bit, would this work?

Author: Yes
